File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        # Mapping from digits to corresponding letters
        phone_map = {
            '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
            '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
        }

        # List to hold the result combinations
        result = []

        # Backtracking function
        def backtrack(index, current_combination):
            # If the current combination is complete
            if index == len(digits):
                result.append(current_combination)
                return
            
            # Get the letters that the current digit can represent
            current_digit = digits[index]
            for letter in phone_map[current_digit]:
                # Append the current letter and move to the next digit
                backtrack(index + 1, current_combination + letter)

        # Start the backtracking with the initial index and empty combination
        backtrack(0, "")
        
        return result
        

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:

    void f(int ind, string curr, string &digits, map<int, string> &m, vector<string> &ans){

        if(ind == digits.size()){ // base case 
            ans.push_back(curr);
            return;
        }

        int num = digits[ind] - '0';
        string letters = m[num];

        for(auto &letter : letters){

            curr.push_back(letter);
            f(ind + 1, curr, digits, m, ans);
            curr.pop_back();
        }

    }
    vector<string> letterCombinations(string digits) {
        
        map<int, string> m = { // map 
            {1, ""},
            {2, "abc"},
            {3, "def"},
            {4, "ghi"},
            {5, "jkl"},
            {6, "mno"},
            {7, "pqrs"},
            {8, "tuv"},
            {9, "wxyz"}
        };

        vector<string> ans; // ans
        if(digits.empty()) return ans;

        f(0, "", digits, m, ans);
        return ans;  
    }
};